### ISSUE TITLE
Circular references are not supported, take two

### ISSUE THREAD

**Issue by Joelius300** [dotnet/aspnetcore#20525]
As correctly mentioned in #18628, blazor currently doesn't support circular references to be serialized. But I'm very confused as to what happened to that issue. It's clear from the sources that the current implementation uses [`JSON.stringify`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) before actually invoking the C# code and that function throws a `TypeError: cyclic object value` when you pass it an object that contains circular references.

#18628 was closed because this is covered in an external issue (https://github.com/dotnet/runtime/issues/30820) but I don't understand how that's the case. It's true that the serializer can't handle circular references yet but once that issue is fixed and the serializer does have support for that, the error in blazor will persist. That's because the error already happens at `JSON.stringify` which isn't affected by blazor. 

I debugged it manually by building and including the source from [Microsoft.JSInterop.JS](https://github.com/dotnet/aspnetcore/blob/master/src/JSInterop/Microsoft.JSInterop.JS/src/). 
When you step through the code you can see that the error happens at `JSON.stringify` (in my case at [Line 61 / `invokePossibleInstanceMethod`](https://github.com/dotnet/aspnetcore/blob/d513eea853c74c43a22cfa99c227cf53acc063d2/src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts#L61)). This is the case when synchronously invoking C# code from JavaScript with an argument that contains circular references. The error that happens is as follows (in Firefox):

> TypeError: cyclic object value

When you do it the other way around (invoke JavaScript from C# with a return value that contains circular references), the error will be rethrown in C# and the error changes to the following. When you debug it you can observe that it's also thrown at a `JSON.stringify` call.

> WASM: Microsoft.JSInterop.JSException: cyclic object value

From my understanding, the error would be exactly the same if the serializer was able to handle circular references because the exception is thrown in JavaScript at `JSON.stringify` and doesn't ever reach the .NET serializer.

Am I completely missing the point here or are my assumptions correct? I would've commented on the other issue but it's locked.  
If I'm correct, will this be supported once https://github.com/dotnet/runtime/issues/30820 is implemented? I'm in need of a solution right now so I'll have to take a look at [flatted](https://github.com/WebReflection/flatted) or something similar in the meantime.

**Comment by javiercn** [comment]
@Joelius300 thanks for contacting us.

When you invoke a JS function and there's an error it gets marshalled back to C# and rethrown there. That behavior is by design. If you put a try..catch you can catch the exception and react to it, otherwise your .NET code would just hang, which is not acceptable.

**Comment by Joelius300** [comment]
I'm sorry but I don't understand how that's relevant. I'm invoking a C# method from JavaScript, not the other way around. I understand the behaviour you explained but that's for C# code calling JavaScript code right?

I'll try to create a minimal example which shows what I mean; the project I used until now is not suitable to showcase this.

**Comment by Joelius300** [comment]
I've created a simple example you can use to reproduce the issue. I just modified the client side blazor template (latest preview).

Index.razor
```html
@page "/"
<h1>Interop error</h1>

<button onclick="setResultOrError(() => JSON.stringify(cyclicObject))" class="btn btn-primary">Stringify</button>
<button onclick="setResultOrError(() => DotNet.invokeMethod('InteropError_CircularReferences', 'ReturnParameterAsText', cyclicObject))" class="btn btn-primary">Invoke .NET Method</button>
<textarea id="myTextarea" style="display:block" cols="100" rows="20" readonly="readonly"></textarea> 
```

script.js (reference it in index.html)
```JavaScript
var cyclicObject = {
    a: 'asfd',
    b: 10
};
cyclicObject.self = cyclicObject;

function setResultOrError(func) {
    try {
        setTextareaValue(func());
    } catch (e) {
        setTextareaValue(e + '\n\n' + e.stack);
    }
}

function setTextareaValue(value) {
    document.getElementById('myTextarea').value = value;
}
```

ExampleInterop.cs
```csharp
using Microsoft.JSInterop;
using System.Text.Json;

namespace InteropError_CircularReferences
{
    public class ExampleInterop
    {
        [JSInvokable]
        public static string ReturnParameterAsText(JsonElement value) => value.GetRawText();
    }
}
```

As you can see, pressing any of the two buttons causes an error which is displayed in the textarea. It's the same error and it says something along the lines of `TypeError: cyclic object value` (may vary depending on your browser). The reason for that was already pointed out correctly in the other issue. Blazor calls `JSON.stringify` internally thus it's exactly the same error.

In #18628 it was suggested that this is an issue (or rather limitation) within the current `System.Text.Json` serializer and was closed for that reason. But when you debug it, you'll realize that you never reach the C#, the error already happens before (as already mentioned, on [Line 61 in `invokePossibleInstanceMethod`](https://github.com/dotnet/aspnetcore/blob/d513eea853c74c43a22cfa99c227cf53acc063d2/src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts#L61)). So my question is why the other issue was closed. I don't see any correlation to the .NET serializer. I get that the serializer doesn't support circular references yet but what does that have to do with this issue? If the serializer would support it right now the error would be exactly the same as far as I understand.

**Comment by Joelius300** [comment]
My initial write up was rather confusing, I'm sorry. I should've included samples etc. I have now edited the original issue and also added code to reproduce in a separate comment.

@javiercn could you take a look at it again?  
If anything is still unclear, just let me know :)

**Comment by danroth27** [comment]
@mkArtakMSFT After reading through the comments here it sounds like there will be work to do in Blazor to support circular references in JS interop calls even once https://github.com/dotnet/runtime/issues/30820 is implemented. Moving this one out of discussions and into the backlog.

@Joelius300 To help us understand the priority of this, could you please share with us some details about your scenario where you need support for circular references?

**Comment by Joelius300** [comment]
Sure.

I've been working on [a blazor port of Chart.js](https://github.com/mariusmuntean/ChartJs.Blazor). When sending some configuration objects to chart.js, we don't need circular references but chart.js has a lot of circular references internally. When we delegate the chart.js callbacks to C# using `IJSRuntime`, we often get a large object containing multiple circular references. We can't transfer this object directly since it will throw at `JSON.stringify`.

The way I overcame this issue is by manually stringifying the object in JavaScript (here TypeScript) using a replacer with `JSON.stringify` to get rid of all the circular references. I then send that raw string over to C# and manually deserialize it.<sup id="a1">[1](#f1)</sup>  
You could also parse the object again in JavaScript and then sent the object to C#. This will however cause a second stringification within blazor which hurts performance a lot for larger objects.

For our project, this workaround is fine since we only encounter circular references from JavaScript to C# and we can just ignore them.


```typescript
private stringifyObjectIgnoreCircular(object: any) {
    const seen = new WeakSet();
    const replacer = (name, value) => {
        if (
            typeof value === "object"
            && value !== null
            && !(value instanceof Boolean)
            && !(value instanceof Date)
            && !(value instanceof Number)
            && !(value instanceof RegExp)
            && !(value instanceof String)
        ) {
            if (seen.has(value))
                return undefined;

            seen.add(value);
        }

        return value;
    }

    return JSON.stringify(object, replacer);
}
```

**Note:** This workaround doesn't factor in `DotNetObjectReference`s which need to be stringified differently.

That being said, it's not top priority for me currently because I have a workaround. Also, I don't think it makes sense to do anything in this regard before https://github.com/dotnet/runtime/issues/30820 is finalized. However, I think once that feature is implemented, blazor should adapt as quickly as possible otherwise the behaviour will be really weird.

<b id="f1">1</b> We still rely on json.net because of some missing features in Sytem.Text.Json and therefore deserialize the raw string with json.net. Doing it with Sytem.Text.Json is also possible and even required if there are (correctly) stringified DotNetObjectReferences. [↩](#a1)

**Comment by gb1337** [comment]
I have the exact same problem as @Joelius, with the same use case of trying to wrap a JS library in Blazor Webassembly. I need to return complex JS objects to .Net, that include references, some of them circular. I would like the ability to: 1. Detect and ignore the circular references; 2. Set a maximum depth of serialization.

**Comment by Joelius300** [comment]
@gb1337 The depth of serialization issue might get fixed with #12685. Make sure to explain your use-case there, it'll increase the chances of it being implemented soon.

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us.
While this may be a great idea, it is not aligned with our long-term vision to make it part of the framework. For many other ideas which don't belong to the framework we encourage the community to build and ship on their own, contributing to the expanding .NET Ecosystem.

You can learn more about our triage process and how we handle issues by reading our [Triage Process](https://github.com/dotnet/aspnetcore/blob/master/docs/TriageProcess.md) writeup.
